src/layout.rs

```diff
@@ -366,126 +382,139 @@ impl Layout {
-            });
-        }
+    // ensure that all the elements are inside the area
+    for element in &elements {
+        solver.add_constraints(&[
+            element.left() | GE(REQUIRED) | area.left(),
+            element.top() | GE(REQUIRED) | area.top(),
+            element.right() | LE(REQUIRED) | area.right(),
+            element.bottom() | LE(REQUIRED) | area.bottom(),
+        ])?;
     }
+
     match layout.direction {
         Direction::Horizontal => {
+            // ensure the elements touch the top and bottom edges of the area
+            for element in &elements {
+                solver.add_constraints(&[
+                    element.top() | EQ(REQUIRED) | area.top(),
+                    element.bottom() | EQ(REQUIRED) | area.bottom(),
+                ])?;
+            }
+            // ensure there are no gaps between the elements
             for pair in elements.windows(2) {
-                ccs.push((pair[0].x + pair[0].width) | EQ(REQUIRED) | pair[1].x);
+                solver.add_constraint(pair[0].right() | EQ(REQUIRED) | pair[1].left())?;
+            }
+            // ensure the first element touches the left edge of the area
+            if let Some(first) = elements.first() {
+                solver.add_constraint(first.left() | EQ(REQUIRED) | area.left())?;
             }
-            for (i, size) in layout.constraints.iter().enumerate() {
-                ccs.push(elements[i].y | EQ(REQUIRED) | f64::from(dest_area.y));
-                ccs.push(elements[i].height | EQ(REQUIRED) | f64::from(dest_area.height));
-                ccs.push(match *size {
-                    Constraint::Length(v) => elements[i].width | EQ(MEDIUM) | f64::from(v),
-                    Constraint::Percentage(v) => {
-                        elements[i].width | EQ(MEDIUM) | (f64::from(v * dest_area.width) / 100.0)
-                    }
-                    Constraint::Ratio(n, d) => {
-                        elements[i].width
-                            | EQ(MEDIUM)
-                            | (f64::from(dest_area.width) * f64::from(n) / f64::from(d))
-                    }
-                    Constraint::Min(v) => elements[i].width | GE(MEDIUM) | f64::from(v),
-                    Constraint::Max(v) => elements[i].width | LE(MEDIUM) | f64::from(v),
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].width | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
+            // ensure the last element touches the right edge of the area. This constraint may not
+            // be satisfied if the constraints are impossible to satisfy.
+            if layout.expand_to_fill {
+                if let Some(last) = elements.last() {
+                    solver.add_constraint(last.right() | EQ(STRONG) | area.right())?;
```

Comment:
```
solver.add_constraint(last.right() | EQ(STRONG) | area.right())?;
```

Should this be `REQUIRED` instead?

Reply:
Actually yeah, I think you're right (incoming push for fixing this)